src/features/device-state/routes/fleet-state-get-v3.ts

```diff
@@ -0,0 +1,372 @@
+	captureException,
+	handleHttpErrors,
+} from '../../../infra/error-handling';
+import {
+	ConfigurationVarsToLabels,
+	EnvVarList,
+	formatImageLocation,
+	readTransaction,
+	rejectUiConfig,
+	setDefaultConfigVariables,
+	varListInsert,
+} from '../state-get-utils';
+import { Expand } from 'pinejs-client-core';
+
+import { sbvrUtils, errors } from '@balena/pinejs';
+import { Application } from '../../../balena-model';
+const { api } = sbvrUtils;
+const { UnauthorizedError } = errors;
+
+type LocalStateApp = FleetStateV3[string]['apps'][string];
+type ServiceComposition = AnyObject;
+export type FleetStateV3 = {
+	[uuid: string]: {
+		name: string;
+		is_managed_by__device?: string;
+		config?: {
+			[varName: string]: string;
+		};
+		apps: {
+			[uuid: string]: {
+				/**
+				 * @deprecated to be removed in state v4
+				 */
+				id: number;
+				name: string;
+				class: 'fleet' | 'block' | 'app';
+				// is_managed_by__device?: string;
+				is_host?: boolean;
+				releases?: {
+					[uuid: string]: {
+						/**
+						 * @deprecated to be removed in state v4
+						 */
+						id: number;
+						services?: {
+							[name: string]: {
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								id: number;
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								image_id: number;
+								image: string;
+								/**
+								 * Defaults to true if undefined
+								 */
+								running?: boolean;
+								environment: {
+									[varName: string]: string;
+								};
+								labels: {
+									[labelName: string]: string;
+								};
+								contract?: AnyObject;
+								composition?: ServiceComposition;
+							};
+						};
+						volumes?: AnyObject;
+						networks?: AnyObject;
+					};
+				};
+			};
+		};
+	};
+};
+
+const serviceBuildByImage = (
+	fleet: AnyObject,
+	image?: AnyObject,
+): undefined | AnyObject => {
+	if (image == null) {
+		return;
+	}
+
+	let serviceId: number;
+	if (typeof image.is_a_build_of__service === 'object') {
+		serviceId = image.is_a_build_of__service.__id;
+	} else {
+		serviceId = image.is_a_build_of__service;
+	}
+
+	return fleet.application__has__service_name.find(
+		(fleetService: AnyObject) => fleetService.id === serviceId,
+	);
+};
+
+function buildFleetAppFromRelease(
+	fleet: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): NonNullable<LocalStateApp['releases']> {
+	let composition: AnyObject = {};
+	const services: NonNullable<LocalStateApp['releases']>[string]['services'] =
+		{};
+
+	// Parse the composition to forward values to the service entries
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch {
+			composition = {};
+		}
+	}
+	for (const ipr of release.contains__image as AnyObject[]) {
+		// extract the per-image information
+		const image = ipr.image[0];
+		const svc = serviceBuildByImage(fleet, image);
```

Comment:
Any reason this can't use `serviceInstallFromImage` the same as for device-state?

Reply:
serviceInstallFromImage accesses the device 
```
	return _.find(device.service_install, (si) => si.service[0].id === id);
```
I can add a switch that the input is either device | fleet and if the input has `service_install` (for device input) I take it otherwise I check for existence of 'service' (for fleet / application as input)